src/OPTokenClaim.sol

```diff
@@ -54,62 +70,139 @@ contract OPTokenClaim is Ownable {
+        return epochs[epoch].totalEXP;
+    }
+
+    /* ========== USER FUNCTIONS ========== */
+
+    /// subscribe to reward distribution for current epoch
+    function subscribe(address account) public {
         Config memory _config = config;
 
         // epoch 0 is the first
-        uint256 epoch = currentEpoch(_config);
+        uint256 epochNum = _currentEpoch(_config);
+        require(epochNum < _config.maxEpoch, "claims ended");
+
+        EpochInfo storage epoch = epochs[epochNum];
+
+        uint256 expBalance = EXP.balanceOf(account);
+        require(expBalance > 0, "address has no exp");
+
+        // cap at MAX_EXP = 99 EXP
+        if (expBalance > MAX_EXP) {
+            expBalance = MAX_EXP;
+        }
+
+        // only add additional EXP tokens to totalEXP if account already subscribed
+        if (epochToSubscribedEXP[epochNum][account] != 0) {
+            uint256 subscribedEXP = epochToSubscribedEXP[epochNum][account];
+            epochToSubscribedEXP[epochNum][account] = expBalance;
+            unchecked {
+                epoch.totalEXP += uint128(expBalance - subscribedEXP);
+            }
+        } else {
+            epochToSubscribedEXP[epochNum][account] = expBalance;
+            unchecked {
+                epoch.totalEXP += uint128(expBalance);
+                epoch.numAccounts++;
+            }
+        }
+
+        emit Subscribed(account, epochNum, expBalance);
+    }
 
-        require(epoch < _config.maxEpoch, "claim period over");
+    /// claim subscribed OP reward
+    function claimOP(address account) external {
+        Config memory _config = config;
 
-        require(
-            !epochToAddressClaimed[epoch][account],
-            "already claimed for this epoch"
-        );
+        // users claim reward for last epoch, so claims start at epoch 1
+        uint256 epochNum = _currentEpoch(_config);
+        require(epochNum > 0, "claims have not started yet");
 
-        uint256 claimableOP = _calcReward(account);
+        uint256 lastEpochNum;
+        unchecked {
+            lastEpochNum = epochNum - 1;
+        }
+        require(lastEpochNum < _config.maxEpoch, "claims ended");
+
+        // check if subscribed and if already claimed
+        require(epochToSubscribedEXP[lastEpochNum][account] > 0, "didn't subscribe or already claimed");
+
+        uint256 OPReward = _calcReward(account, lastEpochNum);
 
         // set claimed to true
-        epochToAddressClaimed[epoch][account] = true;
+        epochToSubscribedEXP[lastEpochNum][account] = 0;
 
         // transfer OP to account
-        require(
-            OP.transferFrom(treasury, account, claimableOP),
-            "Transfer failed"
-        );
+        require(OP.transferFrom(treasury, account, OPReward), "Transfer failed");
+
+        emit OPClaimed(account, lastEpochNum, OPReward);
 
-        emit OPClaimed(account, epoch, claimableOP);
+        if (epochNum < _config.maxEpoch) {
```

Comment:
this should be above OP.transferFrom, keeping a CEI pattern.

Reply:
in [dcede7a](https://github.com/ethernautdao/ethernaut-op-claim/pull/3/commits/dcede7a0b9eb78f7763e975b22aa517e15855bff)